plugin/trino-hive/src/main/java/io/trino/plugin/hive/CachingDirectoryLister.java

```diff
@@ -94,6 +118,65 @@ private static SchemaTablePrefix parseTableName(String tableName)
         return cachingRemoteIterator(iterator, path);
     }
 
+    @Override
+    public void schemaTableDirectoryRenamed(SchemaTableName oldSchemaTableName, Path oldPath, SchemaTableName newSchemaTableName, Path newPath)
+    {
+        invalidateSubPaths(oldSchemaTableName, oldPath);
+    }
+
+    @Override
+    public void schemaTableDirectoryUpdated(SchemaTableName schemaTableName, Path path)
+    {
+        invalidateSubPaths(schemaTableName, path);
+    }
+
+    @Override
+    public void schemaTableDirectoryDeleted(SchemaTableName schemaTableName, Path path)
+    {
+        invalidateSubPaths(schemaTableName, path);
+    }
+
+    private void invalidateSubPaths(SchemaTableName schemaTableName, Path basePath)
+    {
+        if (tablePrefixes.isEmpty()) {
+            // Caching functionality is not being used
+            return;
+        }
+        CompletableFuture.runAsync(() -> {
```

Comment:
1. we don't use the forkjoinpool common pool, as it gives us little control over concurrency
2. as discussed on slack, invalidation *must* be synchronous, i.e. must complete before writing query completes and initiating user fires up a second query

Reply:
I dropped the locking from the current implementation and will be merging the `EvictableCache` changes from https://github.com/trinodb/trino/pull/10949 once they land on `master`.